src/dataService/tagDataService.ts

```diff
@@ -362,6 +377,44 @@ export class TagDataService {
     );
   }
 
+  /**
+   * Bulk update method for creating and deleting tags associated to a Save.
+   * All SaveIds passed to this function must be valid; if any are not
+   * found in the user's saves, the entire operation will be rolled
+   * back and NotFound payload returned.
+   * If attempting to delete a tag that does not exist,
+   * the method is a "no-op" and will not result in an error.
+   * @param updates payload of deletes and creates for a given saveId
+   * @param timestamp timestamp for when the bulk operation occurred
+   * @returns object containing the saveIds that were updated and any
+   * that were missing; these arrays are mutually exclusive (so if one
+   * has any values, the other will be empty).
+   */
+  public async batchUpdateTags(
+    updates: SaveUpdateTagsInputDb,
+    timestamp: Date
+  ): Promise<{ updated: number[]; missing: string[] }> {
+    // First check if the save exists -- no need to initialize
+    // transaction yet
+    const { deletes, creates } = updates;
+    const saveIds = uniqueArray(
+      [...deletes, ...creates].map((req) => parseInt(req.savedItemId))
+    );
+    const missing = await this.pocketSaveService.checkIdExists(saveIds);
+    // Exit if there are any missing Saves, with the missing ids returned
+    if (missing.length) {
+      return { updated: [], missing: missing.map((id) => id.toString()) };
+    }
+    // Otherwise we can proceed with the write
+    // We won't throw error if trying to delete a nonexistent tag (no-op)
+    await this.db.transaction(async (trx) => {
```

Comment:
this is all a single transction, if i'm reading this correctly (and recalling correctly the PR descriptino, which I'm going to double check after I get through this first review - apologies as you've already explained this I believe)?

It does seem like it could be a rather large single transaction if so. I wonder if we could do some load testing of this, to see what upper limits of db hardware resourcing we could possibly hit. Not enough to warrant pausing this PR, just thinking out loud.

Reply:
Yeah absolutely, we both have concerns about this. I'm wondering if a node limit would be good to enforce? But yes I'm not sure what that limit should be, which we probably need to do empirically...

Reply:
I went ahead and added a 150-node write limit (delete/create) as a starting point for testing.